### ISSUE TITLE
[Inductor] [CPU] Torchbench model lennard_jones performance regression > 10% on ww06.2

### ISSUE THREAD

**Issue by ESI-SYD** [pytorch/pytorch#94267]
### 🐛 Describe the bug

Compare with the [TorchInductor CPU Performance Dashboard](https://github.com/pytorch/pytorch/issues/93531#issuecomment-1414956879) on ww05.4, there is a performance regression on Torchbench model lennard_jones on ww06.2 as bellow:

|  ww06.2   |   |     |   |  ww05.4   |   |     |   |   |     |   |
|  ----  | ----  |  ----  | ----  |  ----  | ----  |  ----  | ----  | ----  |  ----  | ----  |
| batch_size  | speedup | inductor  | eager | batch_size  | speedup | inductor  | eager | speedup ratio | eager ratio  | inductor ratio |
|1	|0.9385	|0.0000632	|5.93132E-05	|1	|1.1339	|0.0000545	|6.17976E-05	|0.83	|1.04	|0.86 |

WW06.2 SW info:

SW	| Nightly commit	| Master/Main commit
-- | -- | --
Pytorch|[6a03ad6](https://github.com/pytorch/pytorch/commit/6a03ad6a34d7c2a31fcaab7a28898f1777baac85)|[1d53123](https://github.com/pytorch/pytorch/commit/1d53123f44e2d5f08e4605af353b7d32b62346ae)
Torchbench|/|[ef39f8c](https://github.com/pytorch/benchmark/commit/ef39f8ca47e288ce7206fe92a0eaac03e85623b8)
torchaudio|[ecc2781](https://github.com/pytorch/audio/commit/ecc2781b5b5c353e85a8ca9f16741cf92dd08344)|[4a037b0](https://github.com/pytorch/audio/commit/4a037b03915c4f6f81407e697fb11a7e7ace27fa)
torchtext|[112d757](https://github.com/pytorch/text/commit/112d757efd2482ea5f60bc23d85c2c09f0cfde61)	| [c7cc5fc](https://github.com/pytorch/text/commit/c7cc5fc1e669f548eef619ae055690831d6cc75e)
torchvision|[ac06efe](https://github.com/pytorch/vision/commit/ac06efed4ad2867c68e6b49ac31b554fcbbaa472)|[35f68a0](https://github.com/pytorch/vision/commit/35f68a09f94b2d7afb3f6adc2ba850216413f28e)
torchdata|[049fb62](https://github.com/pytorch/data/commit/049fb626615e6cd965897af8fea2fa73cecd6a2a)|[c0934b9](https://github.com/pytorch/data/commit/c0934b9afa96458c0f8aa3bef528830835c22195)
dynamo_benchmarks|[6a03ad6](https://github.com/pytorch/pytorch/commit/6a03ad6a34d7c2a31fcaab7a28898f1777baac85)|/

WW05.4 SW info: 

SW	| Nightly commit	| Master/Main commit
-- | -- | --
Pytorch|[2e6952f](https://github.com/pytorch/pytorch/commit/2e6952fa58c8b5d779688d16d466faf95555bbb7)|[79db5bc](https://github.com/pytorch/pytorch/commit/79db5bcc9d3febad00e5a2234b44c7db87defdab)
Torchbench|/|[ef39f8c](https://github.com/pytorch/benchmark/commit/ef39f8ca47e288ce7206fe92a0eaac03e85623b8)
torchaudio|[ecc2781](https://github.com/pytorch/audio/commit/ecc2781b5b5c353e85a8ca9f16741cf92dd08344)|[4a037b0](https://github.com/pytorch/audio/commit/4a037b03915c4f6f81407e697fb11a7e7ace27fa)
torchtext|[112d757](https://github.com/pytorch/text/commit/112d757efd2482ea5f60bc23d85c2c09f0cfde61)	| [c7cc5fc](https://github.com/pytorch/text/commit/c7cc5fc1e669f548eef619ae055690831d6cc75e)
torchvision|[ac06efe](https://github.com/pytorch/vision/commit/ac06efed4ad2867c68e6b49ac31b554fcbbaa472)|[35f68a0](https://github.com/pytorch/vision/commit/35f68a09f94b2d7afb3f6adc2ba850216413f28e)
torchdata|[049fb62](https://github.com/pytorch/data/commit/049fb626615e6cd965897af8fea2fa73cecd6a2a)|[c0934b9](https://github.com/pytorch/data/commit/c0934b9afa96458c0f8aa3bef528830835c22195)
dynamo_benchmarks|[b2690c3](https://github.com/pytorch/pytorch/commit/b2690c3ceae36fa6681a0c7cedcc8db7f5d9814a)|/

### Versions

Minified repro

```
python -m torch.backends.xeon.run_cpu --core_list 0 --ncores_per_instance 1 benchmarks/dynamo/torchbench.py --performance --float32 -dcpu -n50 --inductor  --no-skip --dashboard --only lennard_jones  --cold_start_latency --batch_size 1 --threads 1

```

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh

**Comment by chunyuan-w** [comment]
The original behavior of this model: we will have in place buffer reuse.
Starting from https://github.com/pytorch/pytorch/pull/91575, the in-place buffer reuse is no longer there (the behavior of WW05.4).
https://github.com/pytorch/pytorch/pull/92277 has added the in-place buffer reuse back (WW06.2) to align with the original behavior.

For this model, in single batch single thread case, it shows that in-place buffer reuse actually makes the performance worse due to the overhead of `as_strided`.


### The profiling result (for 500 iterations):
**Without in-place buffer**
```
------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                    Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
        mkl::_mkl_linear        96.87%       3.279ms        99.79%       3.378ms       1.351us          2500  
             aten::empty         2.92%      99.000us         2.92%      99.000us       0.040us          2500  
     aten::empty_strided         0.21%       7.000us         0.21%       7.000us       0.014us           500  
    graph_0_kernel_cpp_0         0.00%       0.000us         0.00%       0.000us       0.000us           500  
    graph_0_kernel_cpp_1         0.00%       0.000us         0.00%       0.000us       0.000us           500  
    graph_0_kernel_cpp_2         0.00%       0.000us         0.00%       0.000us       0.000us           500  
    graph_0_kernel_cpp_3         0.00%       0.000us         0.00%       0.000us       0.000us           500  
------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 3.385ms
```

**With in-place buffer**
```
------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                    Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
        mkl::_mkl_linear        84.53%       3.519ms        87.08%       3.625ms       1.450us          2500  
        aten::as_strided        12.92%     538.000us        12.92%     538.000us       0.269us          2000  
             aten::empty         2.55%     106.000us         2.55%     106.000us       0.042us          2500  
    graph_0_kernel_cpp_0         0.00%       0.000us         0.00%       0.000us       0.000us           500  
    graph_0_kernel_cpp_1         0.00%       0.000us         0.00%       0.000us       0.000us           500  
    graph_0_kernel_cpp_2         0.00%       0.000us         0.00%       0.000us       0.000us           500  
    graph_0_kernel_cpp_3         0.00%       0.000us         0.00%       0.000us       0.000us           500  
------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 4.163ms
```

For each iteration,
- without in-place buffer, we will have 1 call of `empty_strided`.
- with in-place buffer, we will have 4 call of `as_strided`.

From the profiling result, the performance gap mainly comes from the gap between time of `empty_strided` and `as_strided`.

### Generated code
**The code without in-place buffer** (which has better performance):
```python
kernel_cpp_0 = async_compile.cpp('''
#include "/tmp/torchinductor_chunyuan/77/c7773nj5pwikpmm2pwa62rcudlf7p3if7eyqb5k4sjsvewwje4le.h"
extern "C" void kernel(const float* __restrict__ in_ptr0,
                       float* __restrict__ out_ptr0)
{
    {
        for(long i0=0; i0<1; i0+=1)
        {
            auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + 16*i0);
            auto tmp1 = decltype(tmp0)(2) / (decltype(tmp0)(1) + (decltype(tmp0)(-2) * tmp0).exp()) - decltype(tmp0)(1);
            tmp1.store(out_ptr0 + 16*i0);
        }
        #pragma omp simd simdlen(8) 
        for(long i0=16; i0<16; i0+=1)
        {
            auto tmp0 = in_ptr0[i0];
            auto tmp1 = std::tanh(tmp0);
            out_ptr0[i0] = tmp1;
        }
    }
}
''')


kernel_cpp_1 = async_compile.cpp('''
#include "/tmp/torchinductor_chunyuan/77/c7773nj5pwikpmm2pwa62rcudlf7p3if7eyqb5k4sjsvewwje4le.h"
extern "C" void kernel(const float* __restrict__ in_ptr0,
                       float* __restrict__ out_ptr0)
{
    {
        for(long i0=0; i0<1; i0+=1)
        {
            auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + 16*i0);
            auto tmp1 = decltype(tmp0)(2) / (decltype(tmp0)(1) + (decltype(tmp0)(-2) * tmp0).exp()) - decltype(tmp0)(1);
            tmp1.store(out_ptr0 + 16*i0);
        }
        #pragma omp simd simdlen(8) 
        for(long i0=16; i0<16; i0+=1)
        {
            auto tmp0 = in_ptr0[i0];
            auto tmp1 = std::tanh(tmp0);
            out_ptr0[i0] = tmp1;
        }
    }
}
''')


kernel_cpp_2 = async_compile.cpp('''
#include "/tmp/torchinductor_chunyuan/77/c7773nj5pwikpmm2pwa62rcudlf7p3if7eyqb5k4sjsvewwje4le.h"
extern "C" void kernel(const float* __restrict__ in_ptr0,
                       float* __restrict__ out_ptr0)
{
    {
        for(long i0=0; i0<1; i0+=1)
        {
            auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + 16*i0);
            auto tmp1 = decltype(tmp0)(2) / (decltype(tmp0)(1) + (decltype(tmp0)(-2) * tmp0).exp()) - decltype(tmp0)(1);
            tmp1.store(out_ptr0 + 16*i0);
        }
        #pragma omp simd simdlen(8) 
        for(long i0=16; i0<16; i0+=1)
        {
            auto tmp0 = in_ptr0[i0];
            auto tmp1 = std::tanh(tmp0);
            out_ptr0[i0] = tmp1;
        }
    }
}
''')


kernel_cpp_3 = async_compile.cpp('''
#include "/tmp/torchinductor_chunyuan/77/c7773nj5pwikpmm2pwa62rcudlf7p3if7eyqb5k4sjsvewwje4le.h"
extern "C" void kernel(const float* __restrict__ in_ptr0,
                       float* __restrict__ out_ptr0)
{
    {
        for(long i0=0; i0<1; i0+=1)
        {
            auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + 16*i0);
            auto tmp1 = decltype(tmp0)(2) / (decltype(tmp0)(1) + (decltype(tmp0)(-2) * tmp0).exp()) - decltype(tmp0)(1);
            tmp1.store(out_ptr0 + 16*i0);
        }
        #pragma omp simd simdlen(8) 
        for(long i0=16; i0<16; i0+=1)
        {
            auto tmp0 = in_ptr0[i0];
            auto tmp1 = std::tanh(tmp0);
            out_ptr0[i0] = tmp1;
        }
    }
}
''')


async_compile.wait(globals())
del async_compile

def call(args):
    arg0_1, arg1_1, arg2_1, arg3_1, arg4_1, arg5_1, arg6_1, arg7_1, arg8_1, arg9_1, arg10_1, arg11_1, arg12_1, arg13_1, arg14_1, arg15_1 = args
    args.clear()
    buf0 = torch.ops.mkl._mkl_linear(arg15_1, arg2_1, arg0_1, arg1_1, 1)
    del arg0_1
    del arg15_1 
    del arg1_1
    del arg2_1
    buf1 = empty_strided((1, 16), (16, 1), device='cpu', dtype=torch.float32)
    kernel_cpp_0(c_void_p(buf0.data_ptr()), c_void_p(buf1.data_ptr()))
    del buf0
    buf2 = torch.ops.mkl._mkl_linear(buf1, arg5_1, arg3_1, arg4_1, 1)
    del arg3_1
    del arg4_1
    del arg5_1
    buf3 = buf1; del buf1  # reuse
    kernel_cpp_1(c_void_p(buf2.data_ptr()), c_void_p(buf3.data_ptr()))
    del buf2
    buf4 = torch.ops.mkl._mkl_linear(buf3, arg8_1, arg6_1, arg7_1, 1)
    del arg6_1
    del arg7_1
    del arg8_1
    buf5 = buf3; del buf3  # reuse
    kernel_cpp_2(c_void_p(buf4.data_ptr()), c_void_p(buf5.data_ptr()))
    del buf4
    buf6 = torch.ops.mkl._mkl_linear(buf5, arg11_1, arg9_1, arg10_1, 1)
    del arg10_1
    del arg11_1
    del arg9_1
    buf7 = buf5; del buf5  # reuse
    kernel_cpp_3(c_void_p(buf6.data_ptr()), c_void_p(buf7.data_ptr()))
    del buf6
    buf8 = torch.ops.mkl._mkl_linear(buf7, arg14_1, arg12_1, arg13_1, 1)
    del arg12_1
    del arg13_1
    del arg14_1
    return (buf8, )
```

**The code with in-place buffer** (worse performance):
```python
kernel_cpp_0 = async_compile.cpp('''
#include "/tmp/torchinductor_chunyuan/77/c7773nj5pwikpmm2pwa62rcudlf7p3if7eyqb5k4sjsvewwje4le.h"
extern "C" void kernel(float* __restrict__ in_out_ptr0)
{
    {
        for(long i0=0; i0<1; i0+=1)
        {
            auto tmp0 = at::vec::Vectorized<float>::loadu(in_out_ptr0 + 16*i0);
            auto tmp1 = decltype(tmp0)(2) / (decltype(tmp0)(1) + (decltype(tmp0)(-2) * tmp0).exp()) - decltype(tmp0)(1);
            tmp1.store(in_out_ptr0 + 16*i0);
        }
        #pragma omp simd simdlen(8) 
        for(long i0=16; i0<16; i0+=1)
        {
            auto tmp0 = in_out_ptr0[i0];
            auto tmp1 = std::tanh(tmp0);
            in_out_ptr0[i0] = tmp1;
        }
    }
}
''')


kernel_cpp_1 = async_compile.cpp('''
#include "/tmp/torchinductor_chunyuan/77/c7773nj5pwikpmm2pwa62rcudlf7p3if7eyqb5k4sjsvewwje4le.h"
extern "C" void kernel(float* __restrict__ in_out_ptr0)
{
    {
        for(long i0=0; i0<1; i0+=1)
        {
            auto tmp0 = at::vec::Vectorized<float>::loadu(in_out_ptr0 + 16*i0);
            auto tmp1 = decltype(tmp0)(2) / (decltype(tmp0)(1) + (decltype(tmp0)(-2) * tmp0).exp()) - decltype(tmp0)(1);
            tmp1.store(in_out_ptr0 + 16*i0);
        }
        #pragma omp simd simdlen(8) 
        for(long i0=16; i0<16; i0+=1)
        {
            auto tmp0 = in_out_ptr0[i0];
            auto tmp1 = std::tanh(tmp0);
            in_out_ptr0[i0] = tmp1;
        }
    }
}
''')


kernel_cpp_2 = async_compile.cpp('''
#include "/tmp/torchinductor_chunyuan/77/c7773nj5pwikpmm2pwa62rcudlf7p3if7eyqb5k4sjsvewwje4le.h"
extern "C" void kernel(float* __restrict__ in_out_ptr0)
{
    {
        for(long i0=0; i0<1; i0+=1)
        {
            auto tmp0 = at::vec::Vectorized<float>::loadu(in_out_ptr0 + 16*i0);
            auto tmp1 = decltype(tmp0)(2) / (decltype(tmp0)(1) + (decltype(tmp0)(-2) * tmp0).exp()) - decltype(tmp0)(1);
            tmp1.store(in_out_ptr0 + 16*i0);
        }
        #pragma omp simd simdlen(8) 
        for(long i0=16; i0<16; i0+=1)
        {
            auto tmp0 = in_out_ptr0[i0];
            auto tmp1 = std::tanh(tmp0);
            in_out_ptr0[i0] = tmp1;
        }
    }
}
''')


kernel_cpp_3 = async_compile.cpp('''
#include "/tmp/torchinductor_chunyuan/77/c7773nj5pwikpmm2pwa62rcudlf7p3if7eyqb5k4sjsvewwje4le.h"
extern "C" void kernel(float* __restrict__ in_out_ptr0)
{
    {
        for(long i0=0; i0<1; i0+=1)
        {
            auto tmp0 = at::vec::Vectorized<float>::loadu(in_out_ptr0 + 16*i0);
            auto tmp1 = decltype(tmp0)(2) / (decltype(tmp0)(1) + (decltype(tmp0)(-2) * tmp0).exp()) - decltype(tmp0)(1);
            tmp1.store(in_out_ptr0 + 16*i0);
        }
        #pragma omp simd simdlen(8) 
        for(long i0=16; i0<16; i0+=1)
        {
            auto tmp0 = in_out_ptr0[i0];
            auto tmp1 = std::tanh(tmp0);
            in_out_ptr0[i0] = tmp1;
        }
    }
}
''')


async_compile.wait(globals())
del async_compile

def call(args):
    arg0_1, arg1_1, arg2_1, arg3_1, arg4_1, arg5_1, arg6_1, arg7_1, arg8_1, arg9_1, arg10_1, arg11_1, arg12_1, arg13_1, arg14_1, arg15_1 = args
    args.clear()
    buf0 = torch.ops.mkl._mkl_linear(arg15_1, arg2_1, arg0_1, arg1_1, 1)
    del arg0_1
    del arg15_1
    del arg1_1
    del arg2_1
    buf1 = as_strided(buf0, (1, 16), (16, 1)); del buf0  # reuse
    kernel_cpp_0(c_void_p(buf1.data_ptr()))
    buf2 = torch.ops.mkl._mkl_linear(buf1, arg5_1, arg3_1, arg4_1, 1)
    del arg3_1
    del arg4_1
    del arg5_1
    del buf1
    buf3 = as_strided(buf2, (1, 16), (16, 1)); del buf2  # reuse
    kernel_cpp_1(c_void_p(buf3.data_ptr()))
    buf4 = torch.ops.mkl._mkl_linear(buf3, arg8_1, arg6_1, arg7_1, 1)
    del arg6_1
    del arg7_1
    del arg8_1
    del buf3
    buf5 = as_strided(buf4, (1, 16), (16, 1)); del buf4  # reuse
    kernel_cpp_2(c_void_p(buf5.data_ptr()))
    buf6 = torch.ops.mkl._mkl_linear(buf5, arg11_1, arg9_1, arg10_1, 1)
    del arg10_1
    del arg11_1
    del arg9_1
    del buf5
    buf7 = as_strided(buf6, (1, 16), (16, 1)); del buf6  # reuse
    kernel_cpp_3(c_void_p(buf7.data_ptr()))
    buf8 = torch.ops.mkl._mkl_linear(buf7, arg14_1, arg12_1, arg13_1, 1)
    del arg12_1
    del arg13_1
    del arg14_1
    return (buf8, )
```

**Comment by chunyuan-w** [comment]
`inplace_buffers` could improve performance in general, but we can't guarantee that it is better in all cases. We've provided an API to turn this function off.
In this case of single-thread single-batch `lennard_jones`, suggest turning off inplace_buffers to get better performance by:
```
from torch._inductor import config as inductor_config
inductor_config.inplace_buffers = False
```

Close this issue as we don't have further plans to fix it.

**Comment by jgong5** [comment]
Just curious why `as_strided` which only manipulates the tensor metadata, no data copy, would take such a long time, compared to `empty_strided` (>70X slower). From the implementation, `as_strided` just has an extra `checkInBoundsForStorage` check, others are common, compared to `empty_strided`? Why slower?

**Comment by chunyuan-w** [comment]
Still not sure why `as_strided` is such slow but I found that `as_strided` is actually not needed in this model: https://github.com/pytorch/pytorch/pull/94481

**Comment by jgong5** [comment]
@chunyuan-w Has this perf regression been fixed by https://github.com/pytorch/pytorch/pull/94481? Of course, the slow `as_strided` perf still needs further investigation but we can open a new issue for it?

**Comment by chunyuan-w** [comment]
> @chunyuan-w Has this perf regression been fixed by #94481? Of course, the slow `as_strided` perf still needs further investigation but we can open a new issue for it?

Yes, it can be fixed by https://github.com/pytorch/pytorch/pull/94481.
Let me close this one and open another issue to track the low perf of `as_strided`.

**Comment by chunyuan-w** [comment]
Opened https://github.com/pytorch/pytorch/issues/94872 to track the `as_strided` performance issue.
